File: iosApp/iosApp/presentation/screens/email_login/EmailLoginViewModel.swift
//
//  EmailLoginViewModel.swift
//  iosApp
//
//  Created by Mostafa Hosseini on 4/23/23.
//  Copyright © 2023 orgName. All rights reserved.
//

import Combine
import Foundation
import KMPNativeCoroutinesCombine
import shared

@MainActor
class EmailLoginViewModel: ObservableObject {
    private let useCase = KoinHelper()
    private var cancellables = Set<AnyCancellable>()

    @Published private(set) var state = EmailLoginState()
    private let eventSubject = PassthroughSubject<EmailEffect, Never>()

    func onEvent(_ event: EmailEvent) {
        switch event {
        case .emailChange(let email):
            updateState(email: email)
        case .continueClicked:
            handleAction {
                registerEmail()
            }
        case .enterBySocials:
            handleAction {
                eventSubject.send(.enterBySocials)
            }
        case .cancelAsyncTasks:
            cancelAsyncTasks()
        }
    }

    private func registerEmail() {
        let emailValidation = useCase.validateEmail(email: state.email)
        updateState(emailError: emailValidation.message)
        if !emailValidation.isSuccess { return }
        updateState(isLoading: true)
        let registerResponse = createFuture(for: useCase.register(email: state.email))
        registerResponse.sink(
            receiveCompletion: { _ in },
            receiveValue: { [self] res in
                updateState(isLoading: false)
                switch res {
                case is ResourceError<RegisteredError>:
                    handleRegisterError(res as! ResourceError<RegisteredError>)
                case is ResourceSuccess<RegisteredSuccess>:
                    guard let res = res as? ResourceSuccess<OtpData> else { return }
                    guard let otp = res.data?.otp else { return }
                    eventSubject.send(.registerSuccessful(otp: otp))
                default: doNothing()
                }
            }
        )
        .store(in: &cancellables)
    }

    private func handleRegisterError(_ res: ResourceError<RegisteredError>) {
        guard let error = res.data?.emailError?.first else {
            eventSubject.send(.error(message: res.message ?? NSLocalizedString("Unknown error. Please try agian", comment: "")))
            return
        }
        updateState(emailError: EmailValidationMessageDynamic(message: error))
    }

    func effect() -> AnyPublisher<EmailEffect, Never> {
        return eventSubject.eraseToAnyPublisher()
    }

    private func updateState(
        email: String? = nil,
        emailError: EmailValidationMessage? = nil,
        isLoading: Bool? = nil,
        isActionEnabled: Bool? = nil
    ) {
        guard let currentState = state.copy() as? EmailLoginState else { return }
        DispatchQueue.main.async { [self] in
            state = state.doCopy(
                email: email ?? currentState.email,
                emailError: emailError ?? currentState.emailError,
                isLoading: isLoading ?? currentState.isLoading,
                isActionsEnabled: isActionEnabled ?? currentState.isActionsEnabled
            )
        }
    }

    private func cancelAsyncTasks() {
        cancellables.forEach { cancellable in
            cancellable.cancel()
        }
    }

    private func doNothing() {}

    @inline(__always)
    private func handleAction(block: () -> Void) {
        updateState(isActionEnabled: false)
        block()
        updateState(isActionEnabled: true)
    }
}

enum EmailEffect {
    case registerSuccessful(otp: String)
    case error(message: String)
    case enterBySocials
}

enum EmailEvent {
    case emailChange(email: String)
    case continueClicked
    case enterBySocials
    case cancelAsyncTasks
}

File: androidApp/src/main/java/com/walletline/android/presentation/screens/auth/email_login/EmailLoginViewModel.kt
package com.walletline.android.presentation.screens.auth.email_login

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.walletline.domain.model.auth.EmailValidationMessage
import com.walletline.domain.model.auth.RegisteredError
import com.walletline.domain.use_case.auth.AuthUseCase
import com.walletline.domain.use_case.validator.ValidateUseCase
import com.walletline.domain.util.Resource
import com.walletline.presentation.screens.auth.email_login.EmailLoginState
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch


class EmailLoginViewModel(
    private val authUseCase: AuthUseCase,
    private val validateUseCase: ValidateUseCase,
) : ViewModel(), EmailLoginContract {

    private val _state = MutableStateFlow(EmailLoginState())
    override val state: StateFlow<EmailLoginState> = _state.asStateFlow()

    private val effectChannel = Channel<EmailLoginContract.Effect>()
    override val effect: Flow<EmailLoginContract.Effect> = effectChannel.receiveAsFlow()

    override fun onEvent(event: EmailLoginContract.Event) {
        when (event) {
            is EmailLoginContract.Event.EmailChange -> _state.update { state -> state.copy(email = event.text) }
            EmailLoginContract.Event.OnContinueClicked -> handleAction { registerEmail() }
            EmailLoginContract.Event.OnEnterBySocialClicked -> handleAction {
                effectChannel.trySend(
                    EmailLoginContract.Effect.EnterBySocial
                )
            }
        }
    }

    private fun registerEmail() {
        val emailValidation = validateUseCase.validateEmail.execute(state.value.email)
        _state.update { state ->
            state.copy(
                emailError = emailValidation.message
            )
        }

        if (!emailValidation.isSuccess) return

        _state.update { state -> state.copy(isLoading = true) }
        viewModelScope.launch {
            authUseCase.register.execute(email = state.value.email)
                .let {
                    _state.update { state -> state.copy(isLoading = false) }
                    when (it) {
                        is Resource.Error -> handleRegisterError(it)
                        is Resource.Success -> {
                            effectChannel.send(
                                EmailLoginContract.Effect.RegisterSuccessful(
                                    otp = it.data.otp,
                                    email = state.value.email
                                )
                            )
                        }
                    }
                }
        }
    }

    private fun handleRegisterError(it: Resource.Error<RegisteredError>) {
        when {
            !it.data?.emailError.isNullOrEmpty() -> {
                _state.update { state ->
                    state.copy(
                        emailError = EmailValidationMessage.Dynamic(it.data?.emailError?.first()!!)
                    )
                }
            }

            else -> effectChannel.trySend(
                EmailLoginContract.Effect.Error(
                    it.message ?: "Unknown error"
                )
            )
        }
    }

    private inline fun handleAction(action: () -> Unit) {
        _state.update { it.copy(isActionsEnabled = false) }
        action()
        _state.update { it.copy(isActionsEnabled = true) }
    }
}